hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/heartbeat/HoodieHeartbeatClient.java

```diff
@@ -0,0 +1,265 @@
+      return lastHeartBeatTime;
+    }
+
+    public void setLastHeartBeatTime(Long lastHeartBeatTime) {
+      this.lastHeartBeatTime = lastHeartBeatTime;
+    }
+
+    public Integer getNumHeartBeatsForCurrentInstantTime() {
+      return numHeartBeatsForCurrentInstantTime;
+    }
+
+    public void setNumHeartBeatsForCurrentInstantTime(Integer numHeartBeatsForCurrentInstantTime) {
+      this.numHeartBeatsForCurrentInstantTime = numHeartBeatsForCurrentInstantTime;
+    }
+
+    public Timer getTimer() {
+      return timer;
+    }
+
+    public void setTimer(Timer timer) {
+      this.timer = timer;
+    }
+  }
+
+  class HeartbeatTask extends TimerTask {
+
+    private final String instantTime;
+
+    HeartbeatTask(String instantTime) {
+      this.instantTime = instantTime;
+    }
+
+    @Override
+    public void run() {
+      updateHeartbeat(instantTime);
+    }
+  }
+
+  /**
+   * Start a new heartbeat for the specified instant. If there is already one running, this will be a NO_OP
+   * @param instantTime
+   */
+  public void start(String instantTime) {
+    LOG.info("Received request to start heartbeat for instant time " + instantTime);
+    Heartbeat heartbeat = instantToHeartbeatMap.get(instantTime);
+    ValidationUtils.checkArgument(heartbeat == null || !heartbeat.isHeartbeatStopped(), "Cannot restart a stopped heartbeat for " + instantTime);
+    if (heartbeat != null && heartbeat.isHeartbeatStarted()) {
+      // heartbeat already started, NO_OP
+      return;
+    } else {
+      Heartbeat newHeartBeat = new Heartbeat();
+      newHeartBeat.setHeartbeatStarted(true);
+      instantToHeartbeatMap.putIfAbsent(instantTime, newHeartBeat);
+      // Ensure heartbeat is generated for the first time with this blocking call.
+      // Since timer submits the task to a thread, no guarantee when that thread will get CPU
+      // cycles to generate the first heartbeat.
+      updateHeartbeat(instantTime);
+      newHeartBeat.getTimer().scheduleAtFixedRate(new HeartbeatTask(instantTime), this.heartbeatIntervalInMs,
+          this.heartbeatIntervalInMs);
+    }
+  }
+
+  /**
+   * Stops the heartbeat for the specified instant.
+   * @param instantTime
+   * @throws HoodieException
+   */
+  public void stop(String instantTime) throws HoodieException {
+    Heartbeat heartbeat = instantToHeartbeatMap.get(instantTime);
+    if (heartbeat != null && heartbeat.isHeartbeatStarted() && !heartbeat.isHeartbeatStopped()) {
+      heartbeat.getTimer().cancel();
+      heartbeat.setHeartbeatStopped(true);
+      LOG.info("Stopping heartbeat for instant " + instantTime);
+    }
+  }
+
+  /**
+   * Stops all heartbeats started via this instance of the client.
+   * @throws HoodieException
+   */
+  public void stop() throws HoodieException {
+    instantToHeartbeatMap.values().stream().forEach(heartbeat -> stop(heartbeat.getInstantTime()));
+  }
+
+  public boolean delete(String instantTime) throws IOException {
```

Comment:
I find this a bit odd, for heartBeat to have a `delete()` method? can this clean up be done as part of `stop()` ? esp given delete() requires a stop first

Reply:
yeah, I agree, we cannot clean up as part of `stop` since we need the heartbeat later when the `cleanFailedWrites` kicks in and then that method does the clean up of heartbeat. I can move this to a utils class if that makes sense.